engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobBatchCollector.java

```diff
@@ -0,0 +1,169 @@
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import java.util.Collection;
+import java.util.function.Predicate;
+import org.agrona.DirectBuffer;
+import org.agrona.ExpandableArrayBuffer;
+import org.agrona.collections.MutableInteger;
+import org.agrona.collections.MutableReference;
+import org.agrona.collections.ObjectHashSet;
+
+/**
+ * Collects jobs to be activated as part of a {@link JobBatchRecord}. Activate-able jobs are read
+ * from the {@link JobState}, resolving and setting their variables from the {@link VariableState},
+ * and added to the given batch record.
+ */
+final class JobBatchCollector {
+  private final ObjectHashSet<DirectBuffer> variableNames = new ObjectHashSet<>();
+
+  private final JobState jobState;
+  private final VariableState variableState;
+  private final Predicate<Integer> canWriteEventOfLength;
+
+  /**
+   * @param jobState the state from which jobs are collected
+   * @param variableState the state from which variables are resolved and collected
+   * @param canWriteEventOfLength a predicate which should return whether the resulting {@link
+   *     TypedRecord} containing the {@link JobBatchRecord} will be writable or not. The predicate
+   *     takes in the size of the record, and should return true if it can write such a record, and
+   *     false otherwise
+   */
+  JobBatchCollector(
+      final JobState jobState,
+      final VariableState variableState,
+      final Predicate<Integer> canWriteEventOfLength) {
+    this.jobState = jobState;
+    this.variableState = variableState;
+    this.canWriteEventOfLength = canWriteEventOfLength;
+  }
+
+  /**
+   * Collects jobs to be added to the given {@code record}. The jobs and their keys are added
+   * directly to the given record.
+   *
+   * <p>This method will fail only if it could not activate anything because the batch would be too
+   * large, but there was at least one job to activate. On failure, it will return that job and its
+   * key. On success, it will return the amount of jobs activated.
+   *
+   * @param record the batch activate command; jobs and their keys will be added directly into it
+   * @return the amount of activated jobs on success, or a job which was too large to activate
+   */
+  Either<LargeJob, Integer> collectJobs(final TypedRecord<JobBatchRecord> record) {
+    final JobBatchRecord value = record.getValue();
+    final ValueArray<JobRecord> jobIterator = value.jobs();
+    final ValueArray<LongValue> jobKeyIterator = value.jobKeys();
+    final Collection<DirectBuffer> requestedVariables = collectVariableNames(value);
+    final var maxActivatedCount = value.getMaxJobsToActivate();
+    final var activatedCount = new MutableInteger(0);
+    final var jobCopyBuffer = new ExpandableArrayBuffer();
+    final var unwritableJob = new MutableReference<LargeJob>();
+
+    jobState.forEachActivatableJobs(
+        value.getTypeBuffer(),
+        (key, jobRecord) -> {
```

Comment:
I would have liked to split this further, as this is quite a big lambda and not that readable in my opinion, but it wasn't really much better due to the amount of large state required for each iteration anyway (plus some of it being mutable).